### ISSUE TITLE
Editing from local serialized Story

### ISSUE THREAD

**Issue by mzorz** [Automattic/stories-android#445]
In order to allow editing, this is how things will work:

Quick summary of convos from this week around how to handle Stories after they’ve been published.

- After a story has been published it will appear in the blog post list
- For the initial release story posts won’t be differentiated from other posts in the list
- Tapping on a story in post list opens the block editor with a static representation of the story block (based on the image block)
- Tapping the edit icon opens the story editor experience
- If the story being opened was created on the same device a user will be able to modify and edit objects in the story*
- If the story was created on a different device we’ll display a message letting the user know that editing functionality for the story is limited.
*To be determined if this will be included for the initial release or shortly thereafter. Need to determine how much extra effort is required.
<img width="1389" alt="Screen Shot 2020-07-18 at 19 46 01" src="https://user-images.githubusercontent.com/6597771/87863242-76ae4600-c92f-11ea-984d-6b3144e90620.png">

#### ~Technical description / flow~ EDIT: THIS CHANGED! Read below 👇🏻
- ~every time a Post is published, we save its serialized Story data and a calculated Post content hash to shared preferences.~
- ~Also check: maybe the background picture / video for each slide will need to exist as local Media, otherwise we can't let the user edit it. Also, I think the content provider URIs provided by apps like Google Photos are only valid for some time, so if they want to edit again the background image / video will no longer be accessible (they can only by adding them to the app again).~
- ~when the user taps on a story post list item (how we detect this has as story is for now by inspection) we check if there's a matching serialized story in shared prefs for this Post. If there is such a thing, we inflate the Story and load it on the Story Creator. Only do this if the hash calculated on the Post content results in the same hash, because we can't possibly allow editing on a Post that has been edited.~
- ~if not, we create a new Story with the background slides being the same as per in the Post, and show the Dialog that says "Limited Story Editing".~
- ~then again every time they tap UPDATE, the Post gets saved, a hash gets calculated for the Post content and stored along with the serialized Story data.~

Also:
- ~deleting a Post will need to check whether there's shared pref for this post, and also delete that information.~

### NEW definitions for technical implementation after conversations with @aforcier, @bjtitus :
We no longer check the Post hash or make any checks prior to opening a Post. We just let them open posts and edit them freely.
We'll be using the flattened produced `mediaId` as the key to Story slides.

#### SAVING:
- when a user saves a Story, we also save to a local (per device) repository the serialized Story slide (frame) and keep the flattened media's mediaID resulting (this is NOT the background image / video `mediaId`, but the new media resulting from saving in the composer). This ID is the one being kept in the block as well. [key = flattened mediaID -> serialized slide.]
The repository should be something that is easy to use on each platform. Caveat: remember that mediaIds are also dependent on the siteId.

#### EDITING:
- when a user opens a post that contains a story block and taps on the block's tools icon to edit it, we do this:
-> mobile gutenberg passes the `mediaFiles` attribute from the block to the host app through the bridge
-> for each `mediaId` we go and check with our local repository to see if we have a match for this siteId/mediaId.
-> if we have a match **_and_** the original background media as pointed to by the serialized object is also still available, bingo this is a fully editable slide.
-> if we don't, we can still present the flattened background image with its mediaId and warn the user with the dialog ("Limited story editing"). We may want to adjust the wording there because it may have been the case the story was actually fully created on this device, but somehow the original background image for a given slide is now not available (for example, with Google Photos temporal Uris?)

### ToDo checklist
- [x] represent the Story block on mobile gutenberg (#504) 
- [ ]  * (tied to this one above) show video placeholder (a per what the Video block does) in the case the Story contains only videos (just point to the first slide)
- [x] build the bridge to determine the icon tool has been tapped
- [x] now pass the content to the bridge (`mediaFiles` array)
- [x] on WPAndroid, resolve how to inflate the serialized slides 
- [x] prior to this, also save the serialized slide that is easily retrievable with `siteId-mediaId` key representing a unique slide.
- [x] show saving / uploading states for the Story block in mobile gutenberg (i.e. when you edit in the composer then tap on DONE then mobile gutenberg should be able to represent the story block being processed / saved  / uploaded as all of this happens.
- [x] check edge case: have a story with some errored story frames and then try to edit it in mobile gutenberg. Represent the block with failed / errored and when they tap the icon, set the error properly. This means we need to serialize the SaveResult and save if to prefs as well.
- [x] if we have a non-editable frame (prefs hasn't been found or the flattened image is not retrievable anymore) -> show warning dialog
- [x] once the user is done editing, we need to implement a callback on the EditPostActivity and back through the bridge to replace the Story block with the new one the user just created.
- [x] also we need to connect the FrameSaveService's StorySaveResult in the upload bridge to not create a new Post but only use the existing, modified Post.

**Other:**
- ~[ ] ideally we just hide the Story block from the block picker in mobile gutenberg so the only way to insert a new Story in a Post is by means of the current flows in place (FAB for new post/page/story on the main screen, and FAB on the app's Posts list)~ - we're letting users insert and add from the block picker
- [ ] cancel error notification after retry #579
- [ ] editing empty story block: should present the media picker instead of presenting the camera capture mode #593 
- [x] design placeholder for empty Story block (edge case) (see https://github.com/Automattic/jetpack/pull/17539)
- [x] check designs, CSS etc. for actual block appearance
- [ ] ~provide a way to clean old preferences (probably add a ShredPreferences.editor.clear() method call when logging out, and also do some checks when a Post is deleted -> check its media associations -> check if any prefs exist -> remove the prefs). Also, check if media is deleted from the Media library -> check if any prefs exist -> remove the prefs.~ (moved to https://github.com/Automattic/stories-android/issues/627)
- [ ] add tracks (see for example [this comment ](add tracking https://github.com/wordpress-mobile/WordPress-Android/pull/12982/files/8905d3ff87c8c2c769a3ca3a5e48cc29714a9b69#r504254019) and other TODOs)
- [x] move WPAndroid dialog logic to a UseCase / utils class  (see https://github.com/wordpress-mobile/WordPress-Android/pull/12982#discussion_r504235226) - addressed in https://github.com/wordpress-mobile/WordPress-Android/pull/13198
- [ ] #439 move Gutenberg processing code to BlockProcessor

**For Later:**
- [ ] probably migrate SharedPrefs to [Jetpack DataStore](https://developer.android.com/topic/libraries/architecture/datastore) #626

These items are being worked on the following PRs:
- Jetpack PR: https://github.com/Automattic/jetpack/pull/17140
- Related Gutenberg PR: https://github.com/WordPress/gutenberg/pull/25242
- Related Gutenberg-mobile PR: https://github.com/wordpress-mobile/gutenberg-mobile/pull/2611
- Related WPAndroid PR: https://github.com/wordpress-mobile/WordPress-Android/pull/12939
- Related Stories library PR: https://github.com/Automattic/stories-android/pull/519

**Comment by mzorz** [comment]
Closing this one as everything is tracked elsewhere as noted in this issue's description
